README.md
# misaem package

### Introduction

`misaem` is a package to perform linear regression and logistic regression with missing data, under MCAR (Missing completely at random) and MAR (Missing at random) mechanisms. The covariates are assumed to be continuous variables. The methodology implemented is based on maximization of the observed likelihood using EM-types of algorithms. The package includes:

1. Parameters estimation.
2. Estimation of standard deviation for estimated parameters.
3. Model selection procedure based on BIC.

### Installation of package
Now you can install the package **misaem** from CRAN.
```{r}
install.packages("misaem")
 ```
 
### Using the misaem package
Basically,

1. `miss.glm` is the main function performing logistic regression with missing values.
2. `miss.lm` is the main function performing linear regression with missing values.

For more details, You can find the vignette, which illustrate the basic and further usage of misaem package:
```{r}
library(misaem)
vignette('misaem')
 ```

## Reference
Logistic Regression with Missing Covariates
-- Parameter Estimation, Model Selection
and Prediction (2020, Jiang W., Josse J., Lavielle M., TraumaBase Group), [Computational Statistics & Data Analysis](https://doi.org/10.1016/j.csda.2019.106907).


NEWS.md
# misaem 1.0.1.9000

Development version (initiated 2021-04-12).

# misaem 1.0.1

A minor release mainly fixing bugs and typos and also serving for authorized maintainer takeover:

* Fix a bug in `miss.lm.fit`, the definition of p, the number of covariates, is now moved to the beginning of the function. A

* Fix another bug in `miss.lm.fit`, replacing `is(x, "data.frame")` with `!is(x, "matrix")` to make the parameter consistency check also pass for a single observation in x.

* Fix some typos in the descriptions of the functions for linear regression model.

* Move `mice` package from imported to suggested packages.

* Takeover of maintenance from Wei Jiang by Julie Josse, who was co-author of the package from the beginning.

# misaem 1.0.0

A major release mainly update the follows:

* Add linear regression with missing values, see the function `miss.lm`.

* The main function for logistic regression is changed to `miss.glm`.

* The functions mimics the structure of widely used functions `lm` and `glm` for the case without missing values.

* Update misaem tutorial.

* Update README.md.

* Typos fixed.

# misaem 0.9.2

A minor release mainly fixing bug:

* Replace `if (class(obj) == "data.frame")  { ..... }` with `if (is(obj, "data.frame"))  { ..... }`. As within the formal S4 class system, a matrix `obj` could well be of a class that extends matrix, such as "matrix" "array" (length = 2).

# misaem 0.9.1

A minor release mainly fixing bugs and typos:

* Fix a bug in `model_selection`, now it can correctly perform model selection if the full model is the best model.

* In `pred_saem`, two methods for prediction of test set with missingness are provided.

* Fix some typos in the vignettes. The length of codes now fits the page wide of html.

* Delete unused Imports `magrittr` in DESCRIPTION file.

* Change the index of vignettes from 'SAEM' to 'misaem tutorial'.

* Update README.md.


R/miss.lm.fit.R
#' Fitting Linear Regression Model with Missing Values
#'
#' This function is used inside \code{miss.lm} to fit linear regression model with missing values, by EM algorithm.
#' @param x design matrix with missingness \eqn{N \times p}{N * p}.
#' @param y response vector \eqn{N \times 1}{N * 1}.
#' @param control a list of parameters for controlling the fitting process. For \code{miss.lm.fit} this is passed to \code{\link{miss.lm.control}}.
#' @return a list with following components:
#' \item{coefficients}{Estimated \eqn{\beta}{\beta}.}
#' \item{ll}{Observed log-likelihood.}
#' \item{s.resid}{Estimated standard error for residuals.}
#' \item{s.err}{Standard error for estimated parameters.}
#' \item{mu.X}{Estimated \eqn{\mu}{\mu}.}
#' \item{Sig.X}{Estimated \eqn{\Sigma}{\Sigma}.}
#' @import norm MASS
#' @importFrom methods is
#' @examples
#' ## For examples see example(miss.lm)

miss.lm.fit <- function (x, y,
                         control = list())

{
  control <- do.call("miss.lm.control", control)
  maxruns = control$maxruns
  tol_em = control$tol_em
  print_iter = control$print_iter

  xnames <- dimnames(x)[[2L]]
  x = as.matrix(x[,-1])
  p <- ncol(x)
  
  #delete rows completely missing
  if(any(apply(is.na(x),1,sum)==p)){
    i_allNA=which(apply(is.na(x),1,sum)==p)
    x = x[-i_allNA,]
    y = y[-i_allNA]
  }
  if(any((is.na(y))==TRUE)){
    i_YNA=which(is.na(y)==TRUE)
    x = x[-i_YNA,]
    y = y[-i_YNA]
  }

  # Check parameter consistency
  if (!is(x, "matrix")){
    x <- as.matrix(x)
  }
  
  if (sum(sapply(x, is.numeric)) < ncol(x) ||
      sum(sapply(y, is.numeric)) < length(y)) {
    stop("Error: parameters 'x' and 'y' should be numeric.")
  }
  # Estimate extended mean and covariance using EM
  n <- nrow(x)
  p <- ncol(x) + 1
  s <- prelim.norm(cbind(y, x))
  thetahat <- em.norm(s, showits=print_iter, maxits=maxruns, criterion=tol_em)
  pars <- getparam.norm(s, thetahat)
  ll <- loglik.norm(s, thetahat)
  ptm <- Sys.time()
  # Calculate regression estimates
  b.est <- c(pars$mu[1] -
               pars$sigma[1,2:p] %*% solve(pars$sigma[2:p,2:p]) %*% pars$mu[2:p],
             pars$sigma[1,2:p] %*% solve(pars$sigma[2:p,2:p]))
  esigma.est <- as.vector(sqrt(
    pars$sigma[1,1] - b.est[2:p] %*% pars$sigma[2:p,2:p] %*% b.est[2:p]))
  Gram.est <- rbind(rep(0, p), cbind(rep(0, p - 1), pars$sigma[2:p,2:p])) +
    c(1, pars$mu[2:p]) %*% t(c(1, pars$mu[2:p]))
  sdb.est <- sqrt(diag(esigma.est^2 * solve(Gram.est * n)))
  time_run=Sys.time() - ptm


  names(b.est) <- xnames
  names(sdb.est) <- xnames

  list(coefficients = b.est, s.err = sdb.est, s.resid = esigma.est, ll= ll, Sig.X = pars$sigma, mu.X = pars$mu)
}


R/print.miss.lm.R
#' Print miss.lm
#'
#' Print results for class \code{miss.lm}.
#' @param x an object of class "\code{miss.lm}", usually, a result of a call to \code{\link{miss.lm}}.
#' @param digits minimal number of significant digits.
#' @param ... further arguments passed to or from other methods.
#' @return No return value, called for coefficient and standard error estimates print.
#' @examples
#' ## For examples see example(miss.lm)
#' @export


print.miss.lm <- function(x, digits = max(3L, getOption("digits") - 3L), ...)
{
  cat("\nCall:  ",
      paste(deparse(x$call), sep = "\n", collapse = "\n"), "\n\n", sep = "")
  if(length(coef(x))) {
    cat("Coefficients")
    cat(":\n")
    print.default(format(x$coefficients, digits = digits),
                  print.gap = 2, quote = FALSE)
  } else cat("No coefficients\n\n")
  cat("Standard error estimates")
  if(length(x$s.err)) {
    if(is.character(co <- x$contrasts))
      cat("  [contrasts: ",
          apply(cbind(names(co),co), 1L, paste, collapse = "="), "]")
    cat(":\n")
    print.default(format(x$s.err, digits = digits),
                  print.gap = 2, quote = FALSE)
  } else cat("No standard error estimates \n\n")
  cat("Log-likelihood:", format(signif(x$ll, digits)))
  cat("\n")
  invisible(x)
}


R/print.miss.glm.R
#' Print miss.glm
#'
#' Print results for class \code{miss.glm}.
#' @param x an object of class "\code{miss.glm}", usually, a result of a call to \code{\link{miss.glm}}.
#' @param digits minimal number of significant digits.
#' @param ... further arguments passed to or from other methods.
#' @return No return value, called for coefficient and standard error estimates print.
#' @examples
#' ## For examples see example(miss.glm)
#' @export


print.miss.glm <- function(x, digits = max(3L, getOption("digits") - 3L), ...)
{
  cat("\nCall:  ",
      paste(deparse(x$call), sep = "\n", collapse = "\n"), "\n\n", sep = "")
  if(length(coef(x))) {
    cat("Coefficients")
    cat(":\n")
    print.default(format(x$coefficients, digits = digits),
                  print.gap = 2, quote = FALSE)
  } else cat("No coefficients\n\n")
  cat("Standard error estimates")
  if(length(x$s.err)) {
    if(is.character(co <- x$contrasts))
      cat("  [contrasts: ",
          apply(cbind(names(co),co), 1L, paste, collapse = "="), "]")
    cat(":\n")
    print.default(format(x$s.err, digits = digits),
                  print.gap = 2, quote = FALSE)
  } else cat("No standard error estimates \n\n")
  cat("Log-likelihood:", format(signif(x$ll, digits)))
  cat("\n")
  invisible(x)
}


R/predict.miss.lm.R
# Predict Method for miss.lm Fits
#'
#' Prediction on test with missing values for the linear regression model.
#' @param object a fitted object of class inheriting from "miss.lm".
#' @param newdata a data frame in which to look for variables with which to predict. It can contain missing values.
#' @param seed  An integer as a seed set for the random generator.
#' @param ... Further arguments passed to or from other methods.
#' @return
#' \item{pr.y}{The prediction result for linear regression.}
#' @import norm MASS
#' @importFrom methods is
#' @examples
#'
#'
#'
#'
#' # Generate complete data
#' set.seed(1)
#' mu.X <- c(1, 1)
#' Sigma.X <- matrix(c(1, 1, 1, 4), nrow = 2)
#' n <- 50 # train set size
#' p <- 2 # number of covariates
#' X.complete <- matrix(rnorm(n*p), nrow=n)%*%chol(Sigma.X) +
#'               matrix(rep(mu.X,n), nrow=n, byrow = TRUE)
#' b <- c(2, 3, -1)
#' sigma.eps <- 0.25
#' y <- cbind(rep(1, n), X.complete) %*% b +
#'   rnorm(n, 0, sigma.eps)
#'
#' # Add missing values
#' p.miss <- 0.10
#' patterns <- runif(n*p)<p.miss #missing completely at random
#' X.obs <- X.complete
#' X.obs[patterns] <- NA
#' # Estimate regression using EM
#' df.obs = data.frame(y ,X.obs)
#' miss.list = miss.lm(y~., data=df.obs)
#'
#' # Generate new dataset for prediction
#' nt <- 20
#' Xt <- matrix(rnorm(nt*p), nrow=nt)%*%chol(Sigma.X)+
#'   matrix(rep(mu.X,nt), nrow=nt, byrow = TRUE)
#' # Generate missingness in new dataset
#' patterns <- runif(nt*p)<p.miss
#' Xt.obs <- Xt
#' Xt.obs[patterns] <- NA
#'
#' # Prediction with missing values
#' miss.pred = predict(miss.list, data.frame(Xt.obs))
#' print(miss.pred)
#' @export

predict.miss.lm <- function(object, newdata = NULL, seed = NA, ...)
{
  if (!is.na(seed))
    set.seed(seed)

  X.new = newdata
  mu.em = object$mu.X
  sig2.em = object$Sig.X
  beta.em = object$coef


  # Check data consistency
  if (is(X.new, "data.frame")){
    X.new <- as.matrix(X.new)
  }
  if (!is.matrix(X.new)){
    stop("Error: parameter 'X.new' should be either a matrix or a data frame.")
  }
  if (sum(sapply(X.new, is.numeric)) < ncol(X.new)) {
    stop("Error: parameter 'X.new should be numeric'.")
  }
  # Prepare structure
  X.prep <- cbind(rep(NA, nrow(X.new)), X.new)
  X.prep <- t(t(X.prep) - mu.em)
  Inv.Sigma.tmp <- solve(sig2.em)
  # Impute
  X.pred <- t(apply(X.prep, 1, imputeEllP, Inv.Sigma.tmp))
  X.pred <- t(t(X.pred) + mu.em)

  return(pr.y=X.pred[,1])
}


R/print.summary.miss.lm.R
#' Print Summary of miss.lm
#'
#' Print results for class \code{summary.miss.lm}.
#' @param x an object of class "\code{summary.miss.lm}", usually, a result of a call to \code{\link{summary.miss.lm}}.
#' @param digits minimal number of significant digits.
#' @param ... further arguments passed to or from other methods.
#' @return No return value, called for summary print.
#' @examples
#' ## For examples see example(miss.lm)
#' @export
print.summary.miss.lm <-
  function (x, digits = max(3L, getOption("digits") - 3L), ...)
  {
    cat("\nCall:\n",
        paste(deparse(x$call), sep = "\n", collapse = "\n"), "\n\n", sep = "")

    if(length(coef(x))) {
      cat("Coefficients")
      cat(":\n")
      print.default(format(x$coefficients, digits = digits),
                    print.gap = 2, quote = FALSE)
    } else cat("No coefficients\n\n")

    cat("Log-likelihood: ", format(x$loglikelihood, digits = max(4L, digits + 1L)),
        "\n", sep = "")
    cat("\n")
    invisible(x)
  }


R/print.summary.miss.glm.R
#' Print Summary of miss.glm
#'
#' Print results for class \code{summary.miss.glm}.
#' @param x an object of class "\code{summary.miss.glm}", usually, a result of a call to \code{\link{summary.miss.glm}}.
#' @param digits minimal number of significant digits.
#' @param ... further arguments passed to or from other methods.
#' @return No return value, called for summary print.
#' @examples
#' ## For examples see example(miss.glm)
#' @export
print.summary.miss.glm <-
  function (x, digits = max(3L, getOption("digits") - 3L), ...)
  {
    cat("\nCall:\n",
        paste(deparse(x$call), sep = "\n", collapse = "\n"), "\n\n", sep = "")

    if(length(coef(x))) {
      cat("Coefficients")
      cat(":\n")
      print.default(format(x$coefficients, digits = digits),
                    print.gap = 2, quote = FALSE)
    } else cat("No coefficients\n\n")

    cat("Log-likelihood: ", format(x$loglikelihood, digits = max(4L, digits + 1L)),
        "\n", sep = "")
    cat("\n")
    invisible(x)
  }


R/miss.lm.R
#' Statistical Inference for Linear Regression Models with Missing Values
#'
#' This function is used to perform statistical inference for linear regression model with missing values, by algorithm EM.
#' @param formula an object of class "\code{\link[stats]{formula}}" : a symbolic description of the linear regression model to be fitted.
#' @param data an optional data frame containing the variables in the model. If not found in \code{data}, the variables are taken from \code{environment}(\code{formula}), typically the environment from which \code{miss.lm} is called.
#' @param control a list of parameters for controlling the fitting process. For \code{miss.lm.fit} this is passed to \code{\link{miss.lm.control}}.
#' @param ... arguments to be used to form the default control argument if it is not supplied directly.
#' @return An object of class "\code{miss.lm}": a list with following components:
#' \item{coefficients}{Estimated \eqn{\beta}{\beta}.}
#' \item{ll}{Observed log-likelihood.}
#' \item{s.resid}{Estimated standard error for residuals.}
#' \item{s.err}{Standard error for estimated parameters.}
#' \item{mu.X}{Estimated \eqn{\mu}{\mu}.}
#' \item{Sig.X}{Estimated \eqn{\Sigma}{\Sigma}.}
#' \item{call}{the matched call.}
#' \item{formula}{the formula supplied.}
#' @import mvtnorm stats
#' @examples
#' # Generate complete data
#' set.seed(1)
#' mu.X <- c(1, 1)
#' Sigma.X <- matrix(c(1, 1, 1, 4), nrow = 2)
#' n <- 50
#' p <- 2
#' X.complete <- matrix(rnorm(n*p), nrow=n)%*%chol(Sigma.X) +
#'               matrix(rep(mu.X,n), nrow=n, byrow = TRUE)
#' b <- c(2, 3, -1)
#' sigma.eps <- 0.25
#' y <- cbind(rep(1, n), X.complete) %*% b + rnorm(n, 0, sigma.eps)
#'
#' # Add missing values
#' p.miss <- 0.10
#' patterns <- runif(n*p)<p.miss #missing completely at random
#' X.obs <- X.complete
#' X.obs[patterns] <- NA
#'
#' # Estimate regression using EM
#' df.obs = data.frame(y,X.obs)
#' miss.list = miss.lm(y~., data=df.obs)
#' print(miss.list)
#' print(summary(miss.list))
#' summary(miss.list)$coef
#' @export

miss.lm <- function (formula,  data,
                     control = list(...), ...
)
{
  call <- match.call()
  if (missing(data))
    data <- environment(formula)
  mf <- match.call(expand.dots = FALSE)
  m <- match(c("formula", "data"), names(mf), 0L)
  mf <- mf[c(1L, m)]
  mf$drop.unused.levels <- TRUE
  mf$na.action <- na.pass #otherwise X will omit NA
  mf[[1L]] <- quote(stats::model.frame)
  mf <- eval(mf, parent.frame())
  control <- do.call("miss.lm.control", control)
  mt <- attr(mf, "terms")
  Y <- model.response(mf, "any")
  if (length(dim(Y)) == 1L) {
    nm <- rownames(Y)
    dim(Y) <- NULL
    if (!is.null(nm))
      names(Y) <- nm
  }
  X <- if (!is.empty.model(mt))
    model.matrix(mt, mf)
  else matrix(,NROW(Y), 0L)

  fit <- eval(call('miss.lm.fit',
                   x = X, y = Y,
                   control = control))
  fit <- c(fit, list(call = call, formula = formula))
  class(fit) <- "miss.lm"
  return(fit)
}
